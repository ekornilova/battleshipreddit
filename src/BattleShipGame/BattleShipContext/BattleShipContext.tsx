import React, { createContext, useContext, useReducer } from 'react'
import { BattleShipContext as BattleShipContextType } from '../types'
import { defaultBattleShipState } from '../constants'
import { BattleShipReducer, DispatchType } from './BattleShipReducer'

type ContextType = BattleShipContextType & {
  dispatch: DispatchType
}

export const BattleShipContext = createContext<ContextType>({
  ...defaultBattleShipState,
  dispatch: () => null,
})

export const BattleShipContextProvider: React.FC = ({ children }) => {
  const [state, dispatch] = useReducer(
    BattleShipReducer,
    defaultBattleShipState
  )
  return (
    <BattleShipContext.Provider value={{ ...state, dispatch }}>
      {children}
    </BattleShipContext.Provider>
  )
}

export const useBattleShipState: () => ContextType = () =>
  useContext(BattleShipContext)
