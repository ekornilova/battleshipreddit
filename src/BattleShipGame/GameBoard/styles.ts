import styled from 'styled-components'
import { Colors, MOBILE_WIDTH, SizeCell, COUNT_CELLS } from '../constants'

export const MainGameViewWrapper = styled.div`
  height: 100%;
  text-align: center;
  display: flex;
  background-color: ${Colors.backgroundColor};
  justify-content: space-around;
  @media (max-width: ${MOBILE_WIDTH}px) {
    flex-direction: column;
    height: unset;
  }
`
export const ShipDeskWrapper = styled.div`
  flex: 0.5;
  display: flex;
  flex-direction: column;
  margin: 10px;
`
export const RestartBtn = styled.button`
  background-color: ${Colors.cellBorder};
  border: 2px solid ${Colors.cellBorder};
  color: ${Colors.backgroundColor};
  border-radius: 50px;
  font-size: 20px;
  font-weight: bold;
  height: 30px;
`
type TextAttr = {
  isBold?: boolean
  size?: number
}
export const Text = styled.div`
  ${(props: TextAttr) => props.isBold && `font-weight: bold;`}
  font-size: ${(props: TextAttr) => props.size || 16}px;
  align-self: center;
  color: ${Colors.textColor};
`

export const ShipImg = styled.img`
  width: 100px;
  height: 60px;
`
export const TableWrapper = styled.div`
  display: grid;
  grid-template-columns: 40% 20% 30%;
  align-items: center;
  grid-auto-rows: 60px;
  justify-items: center;
  margin-bottom: 5px;
`
export const LineWrapper = styled.div`
  display: flex;
`
export const ShipStateImg = styled.img`
  margin-left: 5px;
  border: 2px solid ${Colors.cellBorder};
  width: 20px;
  height: 20px;
`
export const FieldWrapper = styled.div`
  margin: 10px;
  grid-template-columns: repeat(${COUNT_CELLS}, ${SizeCell.desktop}px);
  grid-template-rows: repeat(${COUNT_CELLS}, ${SizeCell.desktop}px);
  display: grid;
  @media (max-width: ${MOBILE_WIDTH}px) {
    grid-template-columns: repeat(${COUNT_CELLS}, ${SizeCell.mobile}px);
    grid-template-rows: repeat(${COUNT_CELLS}, ${SizeCell.mobile}px);
    justify-content: center;
  }
`
export const ModalMessage = styled.div`
  position: fixed;
  background: ${Colors.backgroundColor};
  z-index: 99;
  top: 40%;
  left: 40%;
  width: 25%;
  height: 90px;
  border: 2px solid ${Colors.cellBorder};
  display: flex;
  flex-direction: column;
  border-radius: 40px;
  align-items: end;
  @media (max-width: ${MOBILE_WIDTH}px) {
    width: 65%;
    left: 15%;
  }
}
`
export const ModalBackGround = styled.div`
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  display: flex;
  position: fixed;
  align-items: center;
  justify-content: center;
  background-color: rgba(0, 0, 0, 0.5);
`
