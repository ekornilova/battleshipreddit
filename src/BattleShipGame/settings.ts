import { LayoutType, ShipType } from './types'

export const layoutData: LayoutType = {
  shipTypes: {
    carrier: { size: 5, count: 2 },
    battleship: { size: 4, count: 1 },
    cruiser: { size: 3, count: 1 },
    submarine: { size: 3, count: 1 },
    destroyer: { size: 2, count: 1 },
  },
  layout: [
    {
      ship: ShipType.carrier,
      positions: [
        [2, 9],
        [3, 9],
        [4, 9],
        [5, 9],
        [6, 9],
      ],
    },
    {
      ship: ShipType.carrier,
      positions: [
        [9, 9],
        [9, 8],
        [9, 7],
        [9, 6],
        [9, 5],
      ],
    },
    {
      ship: ShipType.battleship,
      positions: [
        [5, 2],
        [5, 3],
        [5, 4],
        [5, 5],
      ],
    },
    {
      ship: ShipType.cruiser,
      positions: [
        [8, 1],
        [8, 2],
        [8, 3],
      ],
    },
    {
      ship: ShipType.submarine,
      positions: [
        [3, 0],
        [3, 1],
        [3, 2],
      ],
    },
    {
      ship: ShipType.destroyer,
      positions: [
        [0, 0],
        [1, 0],
      ],
    },
  ],
}
