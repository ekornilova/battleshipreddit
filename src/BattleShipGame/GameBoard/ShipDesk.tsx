import React from 'react'
import { useBattleShipState } from '../BattleShipContext/BattleShipContext'
import { ShipType, ShipState } from '../types'
import {
  ShipImg,
  LineWrapper,
  TableWrapper,
  ShipStateImg,
  Text,
} from './styles'
import carrier from '../assets/carrier.png'
import cruiser from '../assets/cruiser.png'
import destroyer from '../assets/destroyer.png'
import battleship from '../assets/battleship.png'
import submarine from '../assets/submarine.png'
import emptyCell from '../assets/empty_cell.png'
import DeathCell from '../assets/death.png'
import FullDeathCell from '../assets/full_death.png'

const ShipStateSrc: Record<ShipState, string> = {
  [ShipState.notTouched]: emptyCell,
  [ShipState.hit]: DeathCell,
  [ShipState.sunk]: FullDeathCell,
}
const ShipSrc: Record<ShipType, string> = {
  [ShipType.carrier]: carrier,
  [ShipType.cruiser]: cruiser,
  [ShipType.battleship]: battleship,
  [ShipType.submarine]: submarine,
  [ShipType.destroyer]: destroyer,
}
const ShipDesk: React.FC = () => {
  const { shipData } = useBattleShipState()
  if (!shipData) {
    return <React.Fragment />
  }
  return (
    <TableWrapper>
      <Text isBold size={20}>
        FLEETS
      </Text>
      <Text isBold size={20}>
        Size
      </Text>
      <Text isBold size={20}>
        State
      </Text>
      {Object.keys(shipData).map((key) => {
        return (
          <React.Fragment key={key}>
            <ShipImg src={ShipSrc[key as ShipType]} alt={key} />
            <Text isBold>{shipData[key as ShipType].size}</Text>
            <LineWrapper>
              {shipData[key as ShipType].ships.map((ship) => {
                return (
                  <ShipStateImg key={ship.id} src={ShipStateSrc[ship.type]} />
                )
              })}
            </LineWrapper>
          </React.Fragment>
        )
      })}
    </TableWrapper>
  )
}
export default ShipDesk
