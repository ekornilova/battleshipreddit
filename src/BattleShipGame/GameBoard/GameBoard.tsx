import React from 'react'
import { useBattleShipState } from '../BattleShipContext/BattleShipContext'
import Cell from '../Cell/Cell'
import { CellType, ActionType } from '../types'
import { FieldWrapper } from './styles'

const GameBoard: React.FC = () => {
  const { field, dispatch } = useBattleShipState()
  const onCellClick = (type: CellType, x: number, y: number) => () => {
    if (type !== CellType.notTouched) {
      return
    }
    dispatch({
      type: ActionType.CHOOSE_CELL,
      payload: {
        x,
        y,
      },
    })
  }
  return (
    <FieldWrapper>
      {field.map((fieldRow, x) => {
        return fieldRow.map((cell, y) => (
          <Cell
            key={`${x}_${y}`}
            onClick={onCellClick(cell.type, x, y)}
            type={cell.type}
          />
        ))
      })}
    </FieldWrapper>
  )
}
export default GameBoard
