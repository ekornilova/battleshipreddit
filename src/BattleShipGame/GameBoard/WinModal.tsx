import React from 'react'
import { Modal } from '../../Components/Modal'
import { Text, RestartBtn, ModalMessage, ModalBackGround } from './styles'

type WinModalProps = {
  onClose: () => void
}
const WinModal: React.FC<WinModalProps> = ({ onClose }) => {
  return (
    <Modal>
      <ModalBackGround>
        <ModalMessage>
          <RestartBtn onClick={onClose}>X</RestartBtn>
          <Text isBold> Congratulations You Win! </Text>
        </ModalMessage>
      </ModalBackGround>
    </Modal>
  )
}
export default WinModal
