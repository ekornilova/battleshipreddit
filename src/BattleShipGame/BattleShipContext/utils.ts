import {
  LayoutType,
  BattleShipContext,
  CellType,
  Cell,
  Ship,
  ShipState,
  ShipStateSettings,
  ShipSetting,
  ShipType,
} from '../types'
import { defaultBattleShipState, COUNT_CELLS } from '../constants'
import { layoutData } from '../settings'

export const isCellWithShip: (cell: Cell) => boolean = (cell: Cell) =>
  cell.shipId !== null
export const setDefaultField = (layoutData: LayoutType): BattleShipContext => {
  const countShips = layoutData.layout.length
  const field: Cell[][] = []
  for (let i = 0; i < COUNT_CELLS; i++) {
    const fieldRow: Cell[] = Array.from(
      {
        length: COUNT_CELLS,
      },
      () => {
        return {
          type: CellType.notTouched,
          shipId: null,
        }
      }
    )
    field.push(fieldRow)
  }
  const shipData: ShipStateSettings = Object.keys(layoutData.shipTypes).reduce(
    (acc, shipKey) => {
      acc[shipKey as ShipType] = {
        ...(layoutData.shipTypes[shipKey as ShipType] as ShipSetting),
        ships: [],
      }
      return acc
    },
    {} as ShipStateSettings
  )
  layoutData.layout.forEach((ship, shipIdx) => {
    shipData[ship.ship].ships.push({
      id: shipIdx,
      type: ShipState.notTouched,
    })
    ship.positions.forEach(([x, y]) => {
      field[x][y].shipId = shipIdx
    })
  })
  return {
    ...defaultBattleShipState,
    shipData,
    layoutData,
    countShips,
    field,
  }
}
export const chooseCell = (
  oldState: BattleShipContext,
  x: number,
  y: number
): BattleShipContext => {
  const field = [...oldState.field]
  let countSunkShips = oldState.countSunkShips
  let shipData = oldState.shipData
  const cell = {
    ...field[x][y],
  }
  if (isCellWithShip(cell)) {
    const shipId = cell.shipId
    cell.type = CellType.hit
    /* eslint-disable-next-line   @typescript-eslint/no-non-null-assertion */
    const positions = oldState.layoutData!.layout[shipId!]!.positions
    if (
      positions.every(
        ([xPos, yPos]) =>
          field[xPos][yPos].type === CellType.hit || (xPos === x && yPos === y)
      )
    ) {
      countSunkShips++
      cell.type = CellType.sunk
      positions.forEach(([xPos, yPos]) => {
        field[xPos][yPos] = {
          ...field[xPos][yPos],
          type: CellType.sunk,
        }
      })
    }
    /* eslint-disable-next-line   @typescript-eslint/no-non-null-assertion */
    const shipType: ShipType = layoutData.layout[shipId!].ship
    if (shipData && shipData[shipType]) {
      const ships: Ship[] = [...shipData[shipType].ships]
      const findShipIdx = ships.findIndex((ship) => shipId === ship.id)
      ships[findShipIdx] = {
        ...ships[findShipIdx],
        type: cell.type === CellType.sunk ? ShipState.sunk : ShipState.hit,
      }
      shipData = {
        ...shipData,
        [shipType]: {
          ...shipData[shipType],
          ships,
        },
      }
    }
  } else {
    cell.type = CellType.miss
  }
  field[x][y] = cell
  return {
    ...oldState,
    shipData,
    field,
    countSunkShips,
  }
}
