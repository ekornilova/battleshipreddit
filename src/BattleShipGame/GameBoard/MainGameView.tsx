import React from 'react'
import { useBattleShipState } from '../BattleShipContext/BattleShipContext'
import { layoutData } from '../settings'
import { ActionType } from '../types'
import { ShipDeskWrapper, MainGameViewWrapper, RestartBtn } from './styles'
import ShipDesk from './ShipDesk'

const MainGameView: React.FC = () => {
  const { dispatch } = useBattleShipState()
  const fillDefaultBoard = React.useCallback(() => {
    dispatch({
      type: ActionType.FILL_DEFAULT_BOARD,
      payload: {
        layoutData,
      },
    })
  }, [dispatch])
  React.useEffect(() => {
    fillDefaultBoard()
  }, [fillDefaultBoard])
  return (
    <MainGameViewWrapper>
      <ShipDeskWrapper>
        <ShipDesk />
        <RestartBtn onClick={fillDefaultBoard}>Restart Game</RestartBtn>
      </ShipDeskWrapper>
    </MainGameViewWrapper>
  )
}
export default MainGameView
