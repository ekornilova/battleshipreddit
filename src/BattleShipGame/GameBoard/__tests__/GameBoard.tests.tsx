import React from 'react'
import * as BattleShipContext from '../../BattleShipContext/BattleShipContext'
import { setDefaultField } from '../../BattleShipContext/utils'
import { layoutData } from '../../settings'
import { cleanup } from '@testing-library/react'
import Enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import GameBoard from '../GameBoard'

Enzyme.configure({ adapter: new Adapter() })

jest.mock('../../BattleShipContext/BattleShipContext')

describe('GameBoard Form Snapshot tests', () => {
  describe('GameBoard', () => {
    beforeEach(() => {
      const defaultState = setDefaultField(layoutData)
      jest
        .spyOn(BattleShipContext, 'useBattleShipState')
        .mockImplementation(() => {
          return {
            ...defaultState,
            dispatch: jest.fn(),
          }
        })
    })
    afterEach(cleanup)

    it(`GameBoard default form`, () => {
      const component = Enzyme.shallow(<GameBoard />)
      expect(component).toMatchSnapshot()
    })
  })
})
