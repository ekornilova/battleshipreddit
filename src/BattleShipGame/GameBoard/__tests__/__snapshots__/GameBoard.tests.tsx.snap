// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameBoard Form Snapshot tests GameBoard GameBoard default form 1`] = `ShallowWrapper {}`;
